src/gamepad/mapping.vala

```diff
@@ -0,0 +1,196 @@
+					int dpad_position = 0;
+					while (dpad_position_2pow > 1) {
+						dpad_position_2pow >>= 1;
+						dpad_position++;
+					}
+					while (dpads.length <= dpad_index) dpads += new Dpad ();
+					dpads[dpad_index].types[dpad_position] = type;
+					dpads[dpad_index].values[dpad_position] = value;
+					break;
+				case 'b':
+					var button = int.parse (real[1:real.length]);
+					while (buttons_type.length <= button) buttons_type += InputType.INVALID;
+					if (buttons_value.length <= button) buttons_value.resize (button + 1);
+					buttons_type[button] = type;
+					buttons_value[button] = value;
+					break;
+				case 'a':
+					var axis = int.parse (real[1:real.length]);
+					while (axes_type.length <= axis) axes_type += InputType.INVALID;
+					if (axes_value.length <= axis) axes_value.resize (axis + 1);
+					axes_type[axis] = type;
+					axes_value[axis] = value;
+					break;
+				}
+			}
+		}
+	}
+
+	public MappedEvent get_dpad_mapping (int dpad_index, int dpad_axis, int dpad_value) {
+		var event = MappedEvent ();
+		int dpad_position;
+		var dpad = dpads[dpad_index];
+		if (dpad_value == 0)
+			dpad_position = (dpad.axisval[dpad_axis] + dpad_axis + 4) % 4;
+		else
+			dpad_position = (dpad_value + dpad_axis + 4) % 4;
+		dpad.axisval[dpad_axis] = dpad_value;
+		event.type = dpad.types[dpad_position];
+		switch (event.type) {
+		case InputType.AXIS:
+			event.axis = (StandardGamepadAxis) dpad.values[dpad_position];
+			break;
+		case InputType.BUTTON:
+			event.button = (StandardGamepadButton) dpad.values[dpad_position];
+			break;
+		}
+		return event;
+	}
+
+	public MappedEvent get_axis_mapping (int axis_number) {
+		var event = MappedEvent ();
+		event.type = axes_type[axis_number];
+		switch (event.type) {
+		case InputType.AXIS:
+			event.axis = (StandardGamepadAxis) axes_value[axis_number];
+			break;
+		case InputType.BUTTON:
+			event.button = (StandardGamepadButton) axes_value[axis_number];
+			break;
+		}
+		return event;
+	}
+
+	public MappedEvent get_button_mapping (int button_number) {
+		var event = MappedEvent ();
+		event.type = buttons_type[button_number];
+		switch (event.type) {
+		case InputType.AXIS:
+			event.axis = (StandardGamepadAxis) buttons_value[button_number];
+			break;
+		case InputType.BUTTON:
+			event.button = (StandardGamepadButton) buttons_value[button_number];
+			break;
+		}
+		return event;
+	}
+
+	public static InputType map_type (string str) {
+		switch (str) {
+		case "leftx":
+		case "lefty":
+		case "rightx":
+		case "righty":
+			return InputType.AXIS;
+		case "a":
+		case "b":
+		case "back":
+		case "dpdown":
+		case "dpleft":
+		case "dpright":
+		case "dpup":
+		case "guide":
+		case "leftshoulder":
+		case "leftstick":
+		case "lefttrigger":
+		case "rightshoulder":
+		case "rightstick":
+		case "righttrigger":
+		case "start":
+		case "x":
+		case "y":
+			return InputType.BUTTON;
+		default:
+			return InputType.INVALID;
+		}
+	}
+
+	public static int map_value (string str) {
```

Comment:
Maybe it's better not to mix return types into int. Instead you could write two methods returning the correct types.